### PR TITLE
Introduce module art.evaluations and implement Security Curve evaluation

### LINKED ISSUE #622: Create module for Evaluations
The new module `art.evaluations` should contain implementations of evaluation procedures of adversarial robustness.

A first example should include an evaluation determining the security curve for a Projected Gradient Descent attack returning accuracy on adversarial and benign data as a function of attack budget `eps`.

### PR DESCRIPTION
**Pull request by beat-buesser**
# Description

This pull request introduces the new module `art.evaluations`. This new module will contain evaluation procedures that go beyond creating adversarial examples and create additional insights into the robustness of machine learning models beyond adversarial accuracy. The ART tools with a main purpose to crate adversarial examples are located in `art.attacks` and new tools in `art.evaluations` should use the tools of the other module like `art.attacks`, `art.defences`, `art.estimators` whenever possible.

This pull request implements as a first evaluation tool the Security Curve analysis based on Projected Gradient Descent that allows automated evaluation of a model's robustness against evasion over the entire range of attack budgets. The Security Curve tool contains methods to produce standard plots of security curves and checks for potential gradient obfuscation.

Fixes #622 

## Type of change

Please check all relevant options.

- [x] Improvement (non-breaking)
- [ ] Bug fix (non-breaking)
- [ ] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: art/evaluations/security_curve/security_curve.ipynb [hunk 1]
```diff
@@ -0,0 +1,374 @@
+{
```
Comment (mathsinn) [inline art/evaluations/security_curve/security_curve.ipynb:1]: Some comments/thoughts on the notebook:

Location: wondering if it should go under notebooks/ ? 

Header cell: I would suggest to rewrite as follows:
```
This notebook demonstrates how to use the ART evaluations module for generating Security Curves. Security Curves have been widely used in the literature to determine and compare the robustness of machine learning models against adversarial examples [1, 2], as well as to identify shortcomings of defences - gradient masking in particular - which could be exploited by an adaptive adversary [3]. Essentially, Security Curves convey the performance of a given model under increasing attack perturbation budgets - commonly in terms of an l_p norm.

[1] A. Madry et al.: Towards Machine Learning Models Resistant to Adversarial Attacks. https://arxiv.org/abs/1706.06083
[2] Y. Dong et al.: Benchmarking Adversarial Robustness on Image Classification. https://arxiv.org/abs/1912.11852
[3] N. Carlini et al.: On Evaluating Adversarial Robustness. https://arxiv.org/abs/1902.06705
```
cell [1]: remove the comment `# from robustness import model_utils, datasets, train, defaults`

cell [4]: typos in "argument" and "except"

cell [9]: "create and" -> "create an"

cell [12]: as per the comment on the security_curve module below: it may be advisable to say that we didn't detect gradient obfuscation, but not categorically saying that the classifier is not obfuscating gradients. 

cell [15]: same typos as in cell [4]

cell [20]: same comment as for cell [12].

**Review by mathsinn — CHANGES_REQUESTED** [review]
Hi @beat-buesser , overall this looks good to me - just left some some minor comments.

File: art/evaluations/security_curve/security_curve.py [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
+# WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+# TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+"""
+This module implements the evaluation of Security Curves.
+
+Examples of Security Curves can be found in Figure 6 of Madry et al., 2019 (https://arxiv.org/abs/1706.06083).
```
[the remark this answers is not in the document]
Comment (beat-buesser) [reply art/evaluations/security_curve/security_curve.py:21]: I agree, I must have accidentally read the year of v4 on arxiv.

File: art/evaluations/security_curve/security_curve.py [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+
+if TYPE_CHECKING:
+    from art.utils import CLASSIFIER_LOSS_GRADIENTS_TYPE
+
+
+class SecurityCurve(Evaluation):
+    """
+    This class implements the evaluation of Security Curves.
+
+    Examples of Security Curves can be found in Figure 6 of Madry et al., 2019 (https://arxiv.org/abs/1706.06083).
```
[the remark this answers is not in the document]
Comment (beat-buesser) [reply art/evaluations/security_curve/security_curve.py:39]: I agree.

File: art/evaluations/security_curve/security_curve.py [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+    def _get_accuracy(y: np.ndarray, y_pred: np.ndarray) -> float:
+        """
+        Calculate accuracy of predicted labels.
+
+        :param y: True labels.
+        :param y_pred: Predicted labels.
+        :return: Accuracy.
+        """
+        num_data = y.shape[0]
+        return np.sum(np.argmax(y, axis=1) == np.argmax(y_pred, axis=1)) / num_data
```
[the remark this answers is not in the document]
Comment (beat-buesser) [reply art/evaluations/security_curve/security_curve.py:179]: I agree.

File: art/evaluations/security_curve/security_curve.py [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+            accuracy_adv = self._get_accuracy(y=y, y_pred=y_pred_adv)
+            self.accuracy_adv_list.append(accuracy_adv)
+
+        # Check gradients for potential obfuscation
+        self._check_gradient(classifier=classifier, x=x, y=y, **kwargs)
+
+        return self.eps_list, self.accuracy_adv_list, self.accuracy
+
+    @property
+    def is_obfuscating_gradients(self) -> bool:
```
[the remark this answers is not in the document]
Comment (beat-buesser) [reply art/evaluations/security_curve/security_curve.py:108]: I agree.

File: art/evaluations/security_curve/security_curve.py [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+
+    def _check_gradient(
+        self,
+        classifier: "CLASSIFIER_LOSS_GRADIENTS_TYPE",
+        x: np.ndarray,
+        y: np.ndarray,
+        **kwargs: Union[str, bool, int, float]
+    ) -> NoReturn:
+        """
+        Check if potential gradient obfuscated can be detected. Projected Gradient Descent with 100 iterations is run
```
[the remark this answers is not in the document]
Comment (beat-buesser) [reply art/evaluations/security_curve/security_curve.py:122]: I agree.

File: art/evaluations/security_curve/security_curve.ipynb (hunk 1 above)
Comment (beat-buesser) [reply art/evaluations/security_curve/security_curve.ipynb]: I think these are great suggestions. I have added them to the notebook and moved it to notebooks/.

**Review by mathsinn — APPROVED** [review]
Thanks a lot @beat-buesser for addressing all the comments. This looks good to me.
